src/NuGet.Core/NuGet.Configuration/PackageNamespaces/SearchTree.cs

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace NuGet.Configuration
+{
+    internal class SearchTree
+    {
+        private readonly SearchNode _root;
+
+        internal SearchTree(PackageNamespacesConfiguration configuration)
+        {
+            if (configuration == null)
+            {
+                throw new ArgumentNullException(nameof(configuration));
+            }
+
+            _root = new SearchNode();
+
+            foreach (string packageSourceKey in configuration.Namespaces.Keys)
+            {
+                foreach (string nugetNamespace in configuration.Namespaces[packageSourceKey])
+                {
+                    Add(packageSourceKey, nugetNamespace);
+                }
+            }
+        }
+
+        private void Add(string packageSourceKey, string namespaceId)
+        {
+            SearchNode currentNode = _root;
+
+            if (namespaceId == null)
+            {
+                throw new ArgumentNullException(nameof(namespaceId));
+            }
+
+            // To prevent from unwanted behaviour.
+            if (namespaceId.Length > PackageNamespacesConfiguration.PackageIdMaxLength)
+            {
+                throw new ArgumentOutOfRangeException(nameof(namespaceId));
+            }
+
+#pragma warning disable CA1308 // Normalize strings to uppercase
+            packageSourceKey = packageSourceKey.ToLowerInvariant().Trim();
+            namespaceId = namespaceId.ToLowerInvariant().Trim();
+#pragma warning restore CA1308 // Normalize strings to uppercase
+
+            for (int i = 0; i < namespaceId.Length; i++)
+            {
+                char c = namespaceId[i];
+
+                if (c == '*')
+                {
+                    // break here since seeing * means end of expression.
+                    currentNode.IsGlobbing = true;
+                    break;
+                }
+
+                if (!currentNode.Children.ContainsKey(c))
+                {
+                    currentNode.Children[c] = new SearchNode();
+                }
+
+                currentNode = currentNode.Children[c];
+            }
+
+            if (string.IsNullOrEmpty(currentNode.NamespaceId))
+            {
+                currentNode.NamespaceId = namespaceId;
+            }
+
+            if (currentNode.PackageSources == null)
+            {
+                currentNode.PackageSources = new HashSet<string>(StringComparer.InvariantCultureIgnoreCase);
```

Comment:
Is a list better suited here? 
HashSet tends to be heavier and given that this is done on a per id basis, it'll happen often enough.

Reply:
Yes, I believe `HashSet` is better for later usage in `PackageDownloader`. Because `List` is O(n) for `Contains`, but `HashSet` is O(1) for `Contains`.

https://github.com/NuGet/NuGet.Client/blob/da32652657dd9a175255b0e0bf2f6bec944b0da1/src/NuGet.Core/NuGet.PackageManagement/PackageDownloader.cs#L129-L130

Reply:
I agree. It would be small set so HashSet doesn't add much value, changed to List.